### DISCUSSION TITLE
Feature Request: a TEXT-Reader

### DISCUSSION THREAD

**Discussion by krahstorfer** [duckdb/duckdb#9571] [Ideas]
### What do you propose?
I propose to add a simple TEXT-Reader. So the content of the text-file (all lines) are in a varchar column. I guess this reader would be way easier to implement as the CSV-reader.

### Why do you want this feature?
I am working a lot with plain text files (e.g. log-files, markdown-files, ...). These files are not really CSV-files, but if I choose an "exotic" column-separator (e.g. "§"), I am normally fine and get the content of the text-file into some DuckDB table for further processing.
With DuckDB 0.9.1 the column separater must be a single byte character now. So my workaround using an "exotic" column-separator no longer works. And all "exotic" single-byte characters are not as exotic as you may think in the first moment, especially if you work with markdown documents ...
With a TEXT-Reader I could get the content of text-files inside DuckDB without any workarounds or pre-processing steps.

**Comment by l1t1** [comment]
https://www.tadviewer.com/ may useful

**Comment by polyrand** [comment]
I'm also very interested in a feature like this. In particular, I would like to replicate the [ClickHouse Template functionality](https://clickhouse.com/docs/en/integrations/data-formats/templates-regexp) by using the `regexp_extract(string, pattern , name_list);` function to generate rows (or structs) based on a regex pattern and (maybe named?) capture groups.

I have been using [this DuckDB snippet](https://duckdbsnippets.com/snippets/11) to achieve this task, but it now breaks with the error:

```
Error: Invalid Input Error: The delimiter option cannot exceed a size of 1 byte.
```

I have started using other single-byte characters to work around the error, but it's become more fragile since I can't always ensure those characters won't exist in the file I'm reading. So, I think it would be useful to have a plain-text reader as a function. Or an extension that can take a regex with named capture groups and a plain text file, and it produces rows.

**Reply by Tishj** [reply]
It sounds like you're looking for this discussion then?
<https://github.com/duckdb/duckdb/discussions/6945>

**Reply by polyrand** [reply]
Thanks for the response, that link was useful.

I was aware of the syntax mentioned in https://github.com/duckdb/duckdb/discussions/6945#discussioncomment-6531020, but if I understand correctly, the group names in the regex are ignored, and the names are assigned based on the array passed as the third argument.

E.g: In this test, I match `foo bar`. `foo` should be `field1`, `bar` should be `field2`.

```sql
-- this works
select regexp_extract('foo bar', '(?P<field1>\w+) (?P<field2>\w+)', ['field1', 'field2']) as m;

-- ┌────────────────────────────────────────┐
-- │                   m                    │
-- │ struct(field1 varchar, field2 varchar) │
-- ├────────────────────────────────────────┤
-- │ {'field1': foo, 'field2': bar}         │
-- └────────────────────────────────────────┘

-- this doesn't work as I would expect (swapping fields)
select regexp_extract('foo bar', '(?P<field1>\w+) (?P<field2>\w+)', ['field2', 'field1']) as m;

-- ┌────────────────────────────────────────┐
-- │                   m                    │
-- │ struct(field2 varchar, field1 varchar) │
-- ├────────────────────────────────────────┤
-- │ {'field2': foo, 'field1': bar}         │
-- └────────────────────────────────────────┘

-- this doesn't return a struct
select regexp_extract('foo bar', '(?P<field1>\w+) (?P<field2>\w+)') as m;

-- ┌─────────┐
-- │    m    │
-- │ varchar │
-- ├─────────┤
-- │ foo bar │
-- └─────────┘

```


In any case, my problem is not really with being able to use named capture groups without explicitly listing them. I just thought it would make writing "parsing templates" easier. My current challenge is knowing how to read a plain text file so that I can parse it using a regex (also assuming that a single "match" can span multiple lines, the plain text files may not necessarily be newline-delimited).

To provide another example. Lets say I have this file:

```
foo__bar
hello__world
```

I can write this command to "convert it to JSON" using ripgrep and query it using DuckDB:

```sh
cat t.txt | \
rg --regexp '(?P<field1>\w+)__(?P<field2>\w+)' --only-matching --replace '{"field1": "$field1", "field2": "$field2"}' | \
duckdb -c "from read_json_auto('/dev/stdin') select field1, field2" 
```

I was trying to come up with a way to substitute `ripgrep` + `read_json_auto('/dev/stdin')` with something like:


```sh
duckdb -c "from read_plain('t.txt', pattern='(?P<field1>\w+)__(?P<field2>\w+)') select field1, field2" 
```

The pattern `'(?P<field1>\w+)__(?P<field2>\w+)'` is just a simplified example.

**Comment by thohan88** [comment]
I agree this would be very useful for logs, html-dumps with json objects, etc. 

I couldn't find it documented anywhere but a workaround is setting `delimiter` and `quote` as `null`:
```
SELECT *
FROM read_csv(<file_name>, 
 delim  = null,
 quote  = null,
 header = false);
```
I could not find a way around the `new_line`-argument so I just use `string_agg(column0)` to collapse the rows to a single line. This works fine for my case, but note that it's does not preserve the ordering so you would need some extra logical if you care about this, e.g. ordering by row_number().

**Reply by Tishj** [reply]
Thanks, as for the ordering problem, you would need to specify the ORDER BY in the aggregate function parameters
Have a look at the railroad diagram here
<https://duckdb.org/docs/sql/aggregates>

**Comment by Maxxen** [comment]
I got an implementation in #10376 that adds a `read_file` table function that partially solves this. The difference is that it doesn't return a VARCHAR column but a BLOB, but you can trivially convert the BLOB into VARCHAR yourself using the `decode` function which will also make sure the data is valid UTF-8 (a requirement for DuckDB strings)
